ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1305,13 +1305,12 @@ private void handleGetResponse(GetResponse getResponse, List<RemoteReplicaInfo>
                     messageInfoList);
               }
 
-              remoteReplicaInfo.setToken(exchangeMetadataResponse.remoteToken);
-              remoteReplicaInfo.setLocalLagFromRemoteInBytes(exchangeMetadataResponse.localLagFromRemoteInBytes);
-              // reset stored metadata response for this replica
-              remoteReplicaInfo.setExchangeMetadataResponse(new ExchangeMetadataResponse(ServerErrorCode.No_Error));
+              if (!validMessageDetectionInputStream.hasInvalidMessages()) {
+                advanceToken(remoteReplicaInfo, exchangeMetadataResponse);
+              }
```

Comment:
There are two ways to do this.
1. Treat MessageFormatException as any other exception and just throw it out in MessageSievingInputStream constructor. This way, whenever replica thread encountered a crc error, it would throw an exception right away and restart the replication.
2. We still treat MessageFormatException differently like this PR. We would deal with other blobs that don't have any error but not advance token so next time we would fetching the same blobs.

Both approaches will perform retry on the MessageFormatException blobs. 

However the second approach would deal with non exception blobs in the first time and later it would fetch these blobs again, which seems to be a better solution.

But both approaches do not solve the infinite retry issue. If a blob is indeed corrupted, then we would fall into infinite retries for this blob. 

We should probably create some alerts for the message format exception if we want to do infinite retry.

Reply:
Sanketh, thanks for reporting this issue. 
Can you help check "replicationMetrics.incrementInvalidMessageError"? We have metrics to record the invalid message. How good or bad it is right now?

Advancing token makes us lose track of the invalid message which is not acceptable.
But if we don't advance it and we'll start and retry from this token every time when the server boots up. If we don't take immediate action on that, we may have bigger and bigger replication lag.

Reply:
My questions and concern is around Ambry Server, not VCR since the code is shared for both Ambry server and VCR.
But I think I have the answer. For Ambry Server, we didn't hit the crc error issues. So the metrics should show zero. So Ambry Server won't have larger replication lag due to the change.